tests/network/mod.rs

```diff
@@ -105,24 +105,20 @@ impl MessageScheduler {
         &self,
         nodes: &BTreeMap<D::NodeId, TestNode<D>>,
     ) -> D::NodeId {
-        match *self {
-            MessageScheduler::First => nodes
-                .iter()
-                .find(|(_, node)| !node.queue.is_empty())
-                .map(|(id, _)| id.clone())
-                .expect("no more messages in queue"),
-            MessageScheduler::Random => {
-                let ids: Vec<D::NodeId> = nodes
-                    .iter()
-                    .filter(|(_, node)| !node.queue.is_empty())
-                    .map(|(id, _)| id.clone())
-                    .collect();
-                rand::thread_rng()
-                    .choose(&ids)
-                    .expect("no more messages in queue")
-                    .clone()
-            }
-        }
+        let mut ids = nodes
+            .iter()
+            .filter(|(_, node)| !node.queue.is_empty())
+            .map(|(id, _)| id.clone());
+        let rand_node = match *self {
+            MessageScheduler::First => rand::thread_rng().gen_weighted_bool(10),
```

Comment:
What if the user implements first non-randomized delivery? Wouldn't DHB fail then?

Reply:
My point is that previously `First` used to work. Is there a fix of the algorithm without changing the test?

Reply:
Yes: What broke it was that in `is_ready` I'm now requiring _N - f_ signatures, which is what we should always expect to be able to reach. I could revert that, but I think it's safer this way. And I think it just exposed what was wrong with the test in the first place.

Reply:
The reason why I changed it in this PR is that before, we had the criterion: _f + 1_ complete `Part`s _and_ the candidate's `Part` must be complete.

Now that the candidates could potentially be _all_ of the new nodes, that criterion would actually be too strong. So I changed it to the strongest criterion that we can always expect to be satisfied, under the assumptions the algorithms are making anyway.

Reply:
OK. This makes the workaround a requirement for validator set changes. How would this map to the networking layer? Would you expect randomization of message delivery there too? This makes it impossible to use `hbbft` in with deterministic delivery such as the previous `First`.

Reply:
In the real world the "first delivery" scenario is unrealistic, and corresponds to a network that would completely block the connection between C and D as long as there are messages sent from A to B. The requirement has not changed:
> The adversary is allowed to control the network as long as every message from a correct node to a correct node eventually arrives.

This requirement was not satisfied by the "first" schedule. That's why the test was wrong, and we were just "lucky" that it passed so far.

Reply:
I agree in general. Now you essentially have two randomized delivery strategies. I think we still need a deterministic one, round-robin for example. I'm not sure what priority that would have given that we move all tests to `net`.

Reply:
I agree, in the long run it would be good to have several different "extreme" delivery strategies (while still satisfying the above requirement). We could easily make this one deterministic, too, by making it do round-robin instead of randomizing in every 10th round.
But I didn't want to spend too much time on implementing delivery strategies for the old test network.

Reply:
I'm getting more convinced about this change but still think that the new validator set feature is not explored enough to be deemed safe and moreover the old test failed on it. This is too big for a single-person review. @c0gent, @mbr, please check if you are happy with @afck's ["real world" argument](https://github.com/poanetwork/hbbft/pull/339/files#r233843400).

Buyer beware: adversaries would stretch the real world boundaries.